include/openmc/tallies/tally_filter.h

```diff
@@ -0,0 +1,88 @@
+#ifndef OPENMC_TALLY_FILTER_H
+#define OPENMC_TALLY_FILTER_H
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "openmc/hdf5_interface.h"
+#include "openmc/particle.h"
+#include "openmc/xml_interface.h"
+
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+extern "C" int32_t n_filters;
+
+class TallyFilterMatch;
+extern std::vector<TallyFilterMatch> filter_matches;
+#pragma omp threadprivate(filter_matches)
+
+class TallyFilter;
+extern std::vector<TallyFilter*> tally_filters;
+
+//==============================================================================
+//! Stores bins and weights for filtered tally events.
+//==============================================================================
+
+class TallyFilterMatch
+{
+public:
+  //int i_bin_;
+  std::vector<int> bins_;
+  std::vector<double> weights_;
+  //bool bins_present_;
+};
+
+//==============================================================================
+//! Modifies tally score events.
+//==============================================================================
+
+class TallyFilter
+{
+public:
+  virtual std::string type() const = 0;
+
+  virtual ~TallyFilter() = 0;
```

Comment:
I decided to go with a pure virtual constructor because I felt like it's the universal way of declaring a class as abstract in C++.  I haven't switched it over in the last round of commits, but not because I disagree; I just need to think about it a bit more.